handler/src/main/java/io/netty/handler/ssl/SniHandler.java

```diff
@@ -313,31 +333,58 @@ public void operationComplete(Future<SslContext> future) throws Exception {
     }
 
     /**
-     * The default implementation will simply call {@link AsyncMapping#map(Object, Promise)} but
-     * users can override this method to implement custom behavior.
+     * The default implementation will simply call {@link SslContextMapping#map(String, Promise)} but users can
+     * override this method to implement custom behavior.
      *
-     * @see AsyncMapping#map(Object, Promise)
+     * NOTE: This method returns a {@code Future<?>} to accommodate the old and new way of doing things without
+     * braking the API. Users who override this method are advised to change their return value from
+     * {@code Future<SslContext>} to {@code Future<Supplier<? extends SslContext>>}.
+     *
+     * @see SslContextMapping#map(String, Promise)
+     * @deprecated the return value will at some point change from {@code Future<SslContext>} to
```

Comment:
WDYT of the following...

Turn `SniHandler` into an `AbstractSniHandler` with an `protected abstract` `select(...)` method.

Let there be a "new" `SniHandler` that extends `AbstractSniHandler` and let it implement the current behavior of using SslContext directly.

Let there be a "real new" `SniWithSupplierHandler` (better name TBD) that extends `AbstractSniHandler` and implements the Supplier based SslContext acquisition. In some ways we could omit this (and the Supplier stuff altogether) because everyone wanting to go this route can built it themselves on top of AbstractSniHandler and we can revisit the issue when more people ask for a concrete impl.